### PR TITLE
Fix bug in Dirichlet.rsample(); add tests

### PR DESCRIPTION
**Pull request by fritzo**
This fixes a bug in `Dirichlet.rsample()` for dim > 2. `Dirichlet.rsample()` was only tested via `Beta.rsample()` using dim=2, so this bug was not caught in previous tests. This PR adds two new tests for `Dirichlet` at n=3. This also numerically stabilizes `torch._C._dirichlet_grad()` by cancelling a factor `(1-x)`.

The bulk of the diff is moving tests around in test_distributions.py. I've done this to keep all `.rsample()` tests together, as they were starting to diverge.

Math and tests reviewed by @martinjankowiak.

### CODE REVIEW

File: torch/distributions/dirichlet.py [hunk 1]
```diff
@@ -5,12 +5,20 @@
...
 def _dirichlet_sample_nograd(alpha):
-    gammas = torch._C._standard_gamma(alpha)
-    return gammas / gammas.sum(-1, True)
+    probs = torch._C._standard_gamma(alpha)
+    probs /= probs.sum(-1, True)
+    eps = _finfo(probs).eps
+    return probs.clamp_(min=eps, max=1 - eps)
+
+
+def _Dirichlet_backward(x, alpha, grad_output):
```
Comment (fritzo) [inline torch/distributions/dirichlet.py]: I've exposed this helper to allow testing.

**Comment by soumith** [comment]
merged  #4572 , so do rebase on top of master whenever you get the chance

**Comment by fritzo** [comment]
cc @apaszke

File: torch/distributions/dirichlet.py [hunk 2]
```diff
@@ -15,6 +15,12 @@ def _dirichlet_sample_nograd(alpha):
     return probs.clamp_(min=eps, max=1 - eps)
 
 
+def _Dirichlet_backward(x, alpha, grad_output):
```
Comment (fritzo) [inline torch/distributions/dirichlet.py]: I've exposed this helper for testing.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
We've had a bug in our travis config and I had to cancel the job. Can you please rebase the PR on top of master, so we can make sure it passes the lint?

File: torch/distributions/dirichlet.py (hunk 2 above)
Comment (apaszke) [reply torch/distributions/dirichlet.py]: Can you please add that comment in the code?

**Pull request merged**